### ISSUE TITLE
Implement way of distributing CUDA-powered application without requirement to have whole CUDA toolkit installed

### ISSUE THREAD

**Issue by rutsky** [inducer/pycuda#70]
Currently I can only compile CUDA file to `cubin` using `pycuda.compiler.compile()` for specific real architecture and then load it using `pycuda.driver.module_from_buffer()`.

AFAIK to distribute CUDA file without compiler/CUDA toolkit I need to either generate `cubin`-s for fixed set of architectures, and/or to generate `ptx` file, that can be assemblied by driver.

It would be nice if PyCUDA will provide some infrastructure for generating, storing and loading `cubin`/`ptx`, so that it would be possible to distribute CUDA-powered applications without CUDA toolkit/compiler.

**Comment by inducer** [comment]
Agreed, I've been thinking about how to best support such a thing. I'd be happy to discuss ideas for a patch.

**Comment by inducer** [comment]
(or take a patch, for that matter.)

**Comment by rutsky** [comment]
With PR #73 I solved the issue in the following way.

I have added custom build step to project setup.py, that compiles all `*.cu` files into `fatbin`'s for real and virtual GPU architectures (i.e. `cubin`'s + PTX's) and saves result `fatbin`'s near `*.cu` files in distribution package. So build system requires to have NVidia CUDA Toolkit, but doesn't require to have CUDA-capable GPU.

In a `.cu` files loading code I have added check: if there is a `fatbin` for the `.cu` file — use it, instead of compiling the `.cu` file. So target system doesn't require to have NVidia CUDA Toolkit, but require to have CUDA-capable GPU with appropriate drivers and CUDA Runtime (that can be legally distributed from NVidia CUDA Toolkit installation).

This solution works for me, but it's not straightforward or well documented for PyCUDA users, and not ready to use "out of the box". Probably you can come up with some utility functions that will ease building and distributing `fatbin`'s in applications.

**Comment by fristedt** [comment]
Any more progress on this? Would be great if we could install pycuda without cuda toolkit.
